### PR TITLE
URL parsing improvements

### LINKED ISSUE #17: Improve .html support via web url input
Currently, 
- We parse and return all the results as a single page. Maybe we break it into pages using `h2` header?
- Output format is different from other supported types

### PR DESCRIPTION
**Pull request by dilithjay**
- Add support for structured parsing of HTML pages
  - Resolves #17 
- Add support for recursive URL parsing: Specify the `depth` keyword argument in the `parse` function to specify how many levels of URLs deep the parsing would be performed. Defaults to `depth = 1` (only parses the specified URL).

**Pending:**
- pdf with hyperlinks: not supported yet. https://github.com/oidlabs-com/Lexoid/issues/26

### CODE REVIEW

File: lexoid/core/utils.py [hunk 1]
```diff
@@ -121,19 +122,214 @@ def download_file(url: str, temp_dir: str) -> str:
...
     Args:
         url (str): The URL of the HTML page.
+        raw (bool): Whether to return raw markdown text or structured data.
 
     Returns:
-        str: The markdown content of the HTML page.
+        Union[str, List[Dict]]: Either raw markdown content or structured data with metadata and content sections.
     """
     response = requests.get(url)
     soup = BeautifulSoup(response.content, "html.parser")
```
Comment (pramitchoudhary) [inline lexoid/core/utils.py:240]: Sometimes decoding errors r thrown,
`Some characters could not be decoded, and were replaced with REPLACEMENT CHARACTER`

Maybe we need to set the encoding explicitly?
```
soup = BeautifulSoup(open(html_path, 'r'),"html.parser",from_encoding="iso-8859-1")
```

Comment (dilithjay) [reply lexoid/core/utils.py]: Fixed.

Comment (dilithjay) [reply lexoid/core/utils.py]: There were also a few bugs in the logic of handling `raw=True`. Should be fixed now.

File: tests/test_parser.py [hunk 2]
```diff
@@ -111,3 +111,18 @@ async def test_url_detection_multi_page_auto_routing(model_type):
...
+@pytest.mark.parametrize("depth", [1, 2])
+async def test_recursive_url_parsing(depth):
+    results = parse("https://example.com/", depth=depth)
+    assert len(results) == depth
+
+
+@pytest.mark.asyncio
+async def test_url_parsing_in_pdf():
+    sample = "examples/inputs/sample_test_doc.pdf"
+    parser_type = "AUTO"
```
Comment (pramitchoudhary) [inline tests/test_parser.py:126]: When testing the following file realized that the first page(home page, https://www.dca.ca.gov/acp/) is not parsed but a different page got parsed, did we miss something?
https://www.dca.ca.gov/acp/pdf_files/lemonlaw_qa.pdf
```
path = "../lemonlaw_qa.pdf"
results = parse(path, parser_type="AUTO", pages_per_split=1, depth=2)
```

Comment (dilithjay) [reply tests/test_parser.py:126]: Seems the URL regex I'm using is faulty. It seems it's extracting the dca.ca.gov from the email address. I updated the regex and added some logic to get it to work. Not sure if there are any other edge cases.

Comment (pramitchoudhary) [reply tests/test_parser.py:126]: Checking ...

Comment (pramitchoudhary) [reply tests/test_parser.py:126]: Working now. Thanks!

**Review by pramitchoudhary — APPROVED** [review]
Thanks @dilithjay 🙏
